### ISSUE TITLE
msgpack API support for shared buffers

### ISSUE THREAD

**Issue by carlosdcastillo** [neovim/neovim#2224]
This started as a conversation on gitter. @justinmk suggested I open an issue.

I'm trying to use Neovim to edit a shared buffer, this functionality is important to me because I'm writing a neovim-mode plugin for Atom. I envision that I can have all of Neovim running perfectly (plugins and everything) and all the UI niceties of Atom, where you can Cmd-F to search and replace. We've
made significant progress:

https://www.youtube.com/watch?v=7TVBcdONEJo

There are two types of actions: Neovim actions and outside actions. 

Example of a Neovim action: the user hits dw in command mode. Example of an outside action: the user opens the Find/Replace window of Atom. Implementing support for outside action has turned out to be incredibly tricky. I think it has to do with the way that the msg pack API is thought out for fully owned buffers, not shared buffers.

In the current status of our plugin, Neovim actions have been handled due to the excellent msgpack API. On outside actions we haven't been able to make any progress.

What right is the situation is that if there is a buffer that has one line and the line is "hello world" and I'm attached to Neovim with 80 columns. Then Neovim will send hello(space)world(space)(space) and 70 spaces more that becomes the atom "buffer" with 70 spaces not in the original file so if the user types A2<esc> it adds a 2 after the d. If the user writes :w then it will be saved perfectly

but if the user goes to Atom File->Save
hello(space)world2(space)(space)(space)(space)(space) gets saved, making a mess
of the file.

Let me emphasize that the root of the problem is that i just get from Neovim location and text that needs to be written there and no information if that text is in the file being edited or if it is to fill in the Neovim grid.

Here are things that would make the msgpack API more friendly to shared buffers that I would love to work on if someone would guide me:
- As I said the big issue that is stopping progress on the Atom side is that there is no way to distinguish a space in the edited from a space that is there for "grid" operations. I think the msgpack API needs to support two types of spaces, ones to fill the empty Neovim grid and ones to represent there is a space in the text file being edited.
- Other big issue that I worked around but it is still an issue is that: for embedding Neovim you need to get the first row that is currently being shown. There should be an API call for this.
- Solving other minor issues such as vim_eval of expand('%:p') or line('$') triggering a full redraw of the screen. See #2070.
- Working on applying changes to the vim buffer efficiently probably by re-visiting #1114, to apply changes to the buffer initiated from Cmd-F Search/Replace initiated from Atom when we get to that point.

(In the text I'm using Atom based Search/Replace as an example of many actions that are initiated from Atom that alter the buffer.)

**Comment by ZyX-I** [comment]
> As I said the big issue that is stopping progress on the Atom side is that there is no way to distinguish a space in the edited from a space that is there for "grid" operations. I think the msgpack API needs to support two types of spaces, ones to fill the empty Neovim grid and ones to represent there is a space in the text file being edited.

1. If you can I would suggest redirecting file save operation to NeoVim. (You in any case must alter `&modified` option.)
2. Even if you can’t you must use buffer contents retrieved with `buffer_get_line_slice` function. I do not think it is ever expected to write screen characters. What will be written with conceal feature enabled (and some characters concealed), `&list` option and `&showbreak` option if you have problems even with non-content spaces?

**Comment by bfredl** [comment]
Somewhat related #1777 , where I tried to share a buffer between a nvim instance and an existing REPL application. (Though this usecase will be simplified significantly by #2076, of course).
It seems this use case would be best served by something "in between" buffer sharing (i e `get/set_line_slice`) and the current ui api, which assumes that the full vim ui will be embedded in a terminal-grid-like. I once thought of something where vim shares the "ui elements", i e slice of rendered lines for some buffer/s/, and separetely, say, the contents of the status line and the commandline (when active) and the current focus/cursor pos, and leaves the "composing" of the elements into a gui layout to the embedding application. 
All that will of course take lots of work, but could be approached in small steps, for instance separate screens for some buffers as I think @tarruda suggested.

**Comment by tarruda** [comment]
> All that will of course take lots of work, but could be approached in steps, for instance separate screens for some buffers as I think @tarruda suggested.

Yes, this will probably require a complete refactor of screen.c, but it will allow the creation of less dumb UIs which take care of window layout and have windows displaying different fonts/sizes for example.

But even multiple screens wont cover @carlosdcastillo use case, which is using Neovim in parallel with atom for editing a file. The ideal for this use case would be a `buffer_attach` API function, analogous for `ui_attach` but operating on a buffer level. After `buffer_attach` is called, the subscriber will receive notifications about line updates, where each payload contains both text and highlighting information for the updated line.

This mechanism for subscribing to buffers is probably much simpler to implement than the multiple screens which I have planned, but certain features implemented at the screen level(fold, breakindent, conceal..) won't be available.

**Comment by tarruda** [comment]
@carlosdcastillo I'm not sure how you are implementing "outside actions", but here's a suggestion: Pass every key to Neovim, and let it forward outside actions via msgpack-rpc requests/notifications back to Atom(a bit like the model-view-controller pattern). For example, say you want to use Atom for finding files, here's a way to do it:

```vim
nnoremap <c-f> call rpcnotify(g:atom_channel_id, 'atom-find-dialog')
```

Where atom would handle `atom-find-dialog` in its msgpack-rpc layer. As @ZyX-I suggested, I also think its better to let Neovim handle file saving.

**Comment by bfredl** [comment]
> The ideal for this would be a buffer_attach API function, analogous to ui_attach but operating on a buffer level. After buffer_attach is called, the subscriber will receive notifications about line updates, where each payload contains both text and highlighting information for the updated line.

:+1: and I would imagine that say the commandline contents could be shared with the same/similar API.
And I agree letting `nvim` handling key events is probably the simplest (in #1777 I did the reverse, and it turned out to be hard to synchronize key focus changes)

**Comment by carlosdcastillo** [comment]
@bfredl The exchange of high level UI-elements (instead of the rendering of these elements) would certainly be awesome. I know this is a fundamental change in the way the API works, but still it would be amazing.

@tarruda I had not thought of that design option. Right now it is if the key combination is not bound to anything on Atom then it gets sent to neovim. I think that since there are Atom plugins that register command that can alter the text the generation of the nvim bindings needs to be dynamic. But it is still doable. I will try to think more about it.

@ZyX-I Are these things (&modified, &list, &showbreak) things I can query from a buffer?

**Comment by ZyX-I** [comment]
@carlosdcastillo What is a point of querying `&list` and `&showbreak`? You can query everything: `getbufvar` in VimL and option-specific API call (don’t know it, but it must exist for `vim.current.buffer.options` in Python API) when you use msgpack.

Note that display options like `&list` are usually *window*-local, not buffer-local. `&showbreak` for some reason is global. Only `&modified` is buffer-local.

**Comment by justinmk** [comment]
> Example of a Neovim action: the user hits dw in command mode. Example of an outside action: the user opens the Find/Replace window of Atom. Implementing support for outside action has turned out to be incredibly tricky.
> ...
> but if the user goes to Atom File->Save
hello(space)world2(space)(space)(space)(space)(space) gets saved, making a mess
of the file.

https://github.com/neovim/neovim/issues/5269 should help with this.

Note that [actualvim](https://github.com/lunixbochs/actualvim) (Sublime 3 plugin that embeds nvim) solves the issue by constantly syncing the full buffer text. #5269 should make that more efficient, but it actually works well already for moderately-sized files (thousands of lines). cc @lunixbochs

**Comment by justinmk** [comment]
Implemented by #7917 / #5269  .
